client/consensus/babe/src/lib.rs

```diff
@@ -347,7 +347,11 @@ impl Config {
 	{
 		trace!(target: "babe", "Getting slot duration");
 
-		let best_block_id = BlockId::Hash(client.usage_info().chain.best_hash);
+		let mut best_block_id = BlockId::Hash(client.usage_info().chain.best_hash);
+		if client.usage_info().chain.finalized_state.is_none() {
+			debug!(target: "babe", "No finalized state not available. Reading config from genesis");
+			best_block_id = BlockId::Hash(client.usage_info().chain.genesis_hash);
```

Comment:
If we are still fast/warp syncing, why do we set something as best block that isn't yet the block? This could break code in a couple of places. 

The syncing code should set something as best block when it has finished syncing the state etc and not before.

Reply:
Not doing so will break code in way more places. First, I should note that only fast sync does that. Warp sync sets the best block only after the full state has been imported. As for fast sync, it downloads the header chains and sets the best header based on the chain selection rule. It needs to verify full header chain before downloading the state, because without warp proofs there's no other way to check if the state is valid. The codebase already supported header-only chain sync, as this is how the light client used to operate before it was removed. The headers as verified and finalised as soon as they are downloaded. In theory we could download full chain and then finalise it at once, but this is too big of a refactoring.

Reply:
Could we then maybe make `best_hash()` a function that is doing this `client.usage_info().chain.finalized_state.is_none()` internally? And also provide some sort of "prospective_best_head"? And `best_hash()` and `prospective_best_head` differ as long as we are doing a fast sync?

Reply:
What would `prospective_best_head` return?
The database contains blocks (headers). Some of them may have state, some may not. State may be not available not just because of the fast sync, but pruning, header-only sync, the light client mode. Although the last was recently removed.
`best_hash` is supposed to return best header hash, not the best available state. Unfortunately the database doe s not keep track of the best available state, only of the best finalized state as it is much more useful. Here I supposed babe could always read configuration from the best avialable finalized state (or genesis if it is unavailable), but I'm not sure of the implications of such a change.

Reply:
Okay